mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitorV2.java

```diff
@@ -0,0 +1,809 @@
+      int type = -1;
+      int subtype = -1;
+      if (mNetworkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) ||
+          mNetworkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI_AWARE)) {
+        type = ConnectivityManager.TYPE_WIFI;
+      } else if (mNetworkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
+        type = ConnectivityManager.TYPE_MOBILE;
+        // To get the subtype we need to make a synchronous ConnectivityManager call
+        // unfortunately. It's recommended to use TelephonyManager.getDataNetworkType()
+        // but that requires an additional permission. Worst case this might be inaccurate
+        // but getting the correct subtype is much much less important than getting the
+        // correct type. Incorrect type could make Envoy Mobile behave like it's offline,
+        // incorrect subtype will just make cellular bandwidth estimates incorrect.
+        NetworkInfo networkInfo = getRawNetworkInfo(network);
+        if (networkInfo != null) {
+          subtype = networkInfo.getSubtype();
+        }
+      } else if (mNetworkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
+        type = ConnectivityManager.TYPE_ETHERNET;
+      } else if (mNetworkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_BLUETOOTH)) {
+        type = ConnectivityManager.TYPE_BLUETOOTH;
+      } else if (mNetworkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) {
+        // Make a synchronous ConnectivityManager call to find underlying network which has a more
+        // useful transport type. crbug.com/1208022
+        NetworkInfo networkInfo = getNetworkInfo(network);
+        type = networkInfo != null ? networkInfo.getType() : ConnectivityManager.TYPE_VPN;
+      }
+      boolean isMetered =
+          !mNetworkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED);
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+        return new NetworkState(
+            true, type, subtype, isMetered,
+            String.valueOf(network.getNetworkHandle()), // NetworkHandle is supported on Android
+                                                        // version M and above
+            mLinkProperties.isPrivateDnsActive(), mLinkProperties.getPrivateDnsServerName());
+      }
+      return new NetworkState(
+          true, type, subtype, isMetered,
+          String.valueOf(network.getNetworkHandle()), // NetworkHandle is supported on Android
+                                                      // version M and above
+          false, "");
+    }
+  }
+
+  // This class gets called back by ConnectivityManager whenever networks come
+  // and go. It gets called back on a special handler thread
+  // ConnectivityManager creates for making the callbacks. The callbacks in
+  // turn post to mLooper where mObserver lives.
+  private class MyNetworkCallback extends NetworkCallback {
```

Comment:
I realize Chromium named this class the same, but `MyNetworkCallback` isn't very descriptive...  how about `NetworkChangesCallback` or something like that?